sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataobject/SnowflakeTableDataObject.scala

```diff
@@ -19,93 +19,85 @@
-import io.smartdatalake.definitions.{SDLSaveMode, SaveModeOptions}
 import io.smartdatalake.definitions.SDLSaveMode.SDLSaveMode
-import io.smartdatalake.util.misc.DataFrameUtil.DfSDL
+import io.smartdatalake.definitions.{SDLSaveMode, SaveModeOptions}
+import io.smartdatalake.smartdatalake.{SnowparkDataFrame, SnowparkStructType, SparkDataFrame, SparkStructType}
 import io.smartdatalake.util.hdfs.PartitionValues
-import io.smartdatalake.util.misc.DataFrameUtil
+import io.smartdatalake.util.misc.DataFrameUtil.DfSDL
 import io.smartdatalake.workflow.ActionPipelineContext
-import io.smartdatalake.workflow.connection.SnowflakeTableConnection
+import io.smartdatalake.workflow.connection.SnowflakeConnection
 import net.snowflake.spark.snowflake.Utils.SNOWFLAKE_SOURCE_NAME
-import org.apache.spark.sql.types.StructType
-import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
 
 /**
  * [[DataObject]] of type SnowflakeTableDataObject.
  * Provides details to access Snowflake tables via an action
+ * Can be used both for interacting with Snowflake through Spark with JDBC,
+ * as well as for actions written in the Snowpark API that run directly on Snowflake
  *
  * @param id           unique name of this data object
- * @param schemaMin    An optional, minimal schema that this DataObject must have to pass schema validation on reading and writing.
  * @param table        Snowflake table to be written by this output
  * @param saveMode     spark [[SDLSaveMode]] to use when writing files, default is "overwrite"
  * @param connectionId The SnowflakeTableConnection to use for the table
  * @param comment      An optional comment to add to the table after writing a DataFrame to it
  * @param metadata     meta data
  */
 case class SnowflakeTableDataObject(override val id: DataObjectId,
-                                    override val schemaMin: Option[StructType] = None,
                                     override var table: Table,
+                                    override val schemaMin: Option[SparkStructType] = None,
                                     saveMode: SDLSaveMode = SDLSaveMode.Overwrite,
                                     connectionId: ConnectionId,
                                     comment: Option[String],
                                     override val metadata: Option[DataObjectMetadata] = None)
                                    (@transient implicit val instanceRegistry: InstanceRegistry)
-  extends TransactionalSparkTableDataObject {
+  extends TransactionalSparkTableDataObject
+    with CanCreateSnowparkDataFrame
+    with CanWriteSnowparkDataFrame {
 
-  /**
-   * Connection defines connection string, credentials and database schema/name
-   */
-  private val connection = getConnection[SnowflakeTableConnection](connectionId)
+  private val connection = getConnection[SnowflakeConnection](connectionId)
+  private var _snowparkSession: Option[Session] = None
```

Comment:
Ah, you create a session per DataObject. But it could be shared by connection & schema, right?
Maybe you should add the schema to the SnowflakeConnection?

Reply:
You are right. I moved the cache down to the connection. A valid schema is needed to create the session, but after that, that schema is ignored, so there really only needs to be a single cached session per connection.